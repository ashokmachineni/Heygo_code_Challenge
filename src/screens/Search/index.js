import React, {useState} from 'react';
import {
  SafeAreaView,
  View,
  ActivityIndicator,
  Keyboard,
  StyleSheet,
} from 'react-native';
import {useDispatch, useSelector} from 'react-redux';
import {search, location} from '../../store/actions';
import Input from './widget/Input';
import List from './widget/List';
import {MAP_SCREEN} from '../../navigator/routes';

const Search = ({navigation}) => {
  const [query, setQuery] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const dispatch = useDispatch();
  const searchData = useSelector(({searchData}) => searchData);

  const onChangeQuery = text => {
    dispatch(search(text));
    setQuery(text);
  };

  const onPressTitle = async id => {
    Keyboard.dismiss();
    setIsLoading(true);
    try {
      const res = await dispatch(location(id));
      res && navigation.navigate(MAP_SCREEN);
    } catch (e) {
      console.log('ERROR::', e);
    } finally {
      setIsLoading(false);
    }
  };
  return (
    <SafeAreaView style={{flex: 1}}>
      <View style={{flex: 1}}>
        <Input value={query} onChange={e => onChangeQuery(e)} />
        {isLoading && (
          <View style={s.loader}>
            <ActivityIndicator size="large" color="#0000ff" />
          </View>
        )}
        <View accessibilityLabel="list" style={{paddingHorizontal: 20}}>
          <List data={searchData} onPressTitle={onPressTitle} />
        </View>
      </View>
    </SafeAreaView>
  );
};

export default Search;

const s = StyleSheet.create({
  loader: {
    position: 'absolute',
    flex: 1,
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    height: '100%',
    justifyContent: 'center',
    backgroundColor: '#0000004D',
  },
});
