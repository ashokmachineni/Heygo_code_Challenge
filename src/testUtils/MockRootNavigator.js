import React from 'react';
import {SEARCH_SCREEN, MAP_SCREEN} from '../navigator/routes';
import {NavigationContainer} from '@react-navigation/native';
import {createStackNavigator} from '@react-navigation/stack';

const Stack = createStackNavigator();

const MockRootNavigator = ({component}) => {
  return (
    <NavigationContainer>
      <Stack.Navigator headerMode="none">
        <Stack.Screen name={SEARCH_SCREEN}>{props => component}</Stack.Screen>
        <Stack.Screen name={MAP_SCREEN}>{props => component}</Stack.Screen>
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default MockRootNavigator;
