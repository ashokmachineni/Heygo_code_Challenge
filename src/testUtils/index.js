import React from 'react';
import {createStore, applyMiddleware} from 'redux';
import {Provider} from 'react-redux';
import {render as renderComponent} from '@testing-library/react-native';
import createMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import reducers from '../store/reducers';
import {mergeDeepRight} from 'ramda';
import MockRootNavigator from './MockRootNavigator';

const defaultOptions = {
  initialState: {
    searchData: [],
    locationData: [],
  },
  api: {},
};

export const renderForTest = (
  Component,
  {initialState, api} = defaultOptions,
) => {
  const mockStore = configureMockStore(api);
  return renderComponent(
    <Provider store={mockStore(initialState)}>
      <MockRootNavigator component={Component} />
    </Provider>,
  );
};

export const configureMockStore =
  api =>
  (initialState = {}) =>
    createMockStore([thunk.withExtraArgument({api})])(
      mergeDeepRight(defaultOptions.initialState, initialState),
    );

export const configureStore = (api, initialState) =>
  createStore(
    reducers,
    initialState,
    applyMiddleware(thunk.withExtraArgument({api})),
  );

export const render = (component, {initialState, api} = defaultOptions) => {
  const store = configureStore(
    {...defaultOptions.api, ...api},
    {...defaultOptions.initialState, ...initialState},
  );

  const Component = renderComponent(
    <Provider store={store}>
      <MockRootNavigator component={component} />
    </Provider>,
  );

  const rerender = () => {
    Component.rerender(
      <Provider store={store}>
        <MockRootNavigator component={component} />
      </Provider>,
    );
  };

  return [Component, store, rerender];
};

export const renderForTestWithStore = render;
