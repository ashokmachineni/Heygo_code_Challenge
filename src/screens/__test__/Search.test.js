import React from 'react';
import Search from '../Search';
import {renderForTest} from '../../testUtils';
import {fireEvent} from '@testing-library/react-native';
let field;
let list;
let onPress;
describe('Search Screen', () => {
  it('Search screen should render correctly', () => {
    const search = renderForTest(<Search />).toJSON();
    expect(search).toMatchSnapshot();
  });
  it('search screen Input', async () => {
    const screen = renderForTest(<Search />);
    field = screen.getByPlaceholderText('Search');
    onPress = screen.getByLabelText('List');
    fireEvent.changeText(field, 'lon');
  });

  it('search result list', async () => {
    const screen = renderForTest(<Search />);
    list = screen.getByLabelText('list');
  });
});
