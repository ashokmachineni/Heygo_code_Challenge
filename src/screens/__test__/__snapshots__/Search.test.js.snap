// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Screen Search screen should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenContainer
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        activityState={2}
        forwardedRef={[Function]}
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "opacity": 1,
            }
          }
        />
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onGestureEnd={[Function]}
          onOpen={[Function]}
          onTransition={[Function]}
          pointerEvents="box-none"
          style={
            Array [
              Object {
                "overflow": undefined,
              },
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
            ]
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            needsOffscreenAlphaCompositing={false}
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              collapsable={false}
              forwardedRef={[Function]}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              style={
                Object {
                  "flex": 1,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": -1,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 5,
                    "top": 0,
                    "width": 3,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <RCTSafeAreaView
                      emulateUnlessSupported={true}
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        accessible={true}
                        focusable={true}
                        onClick={[Function]}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          Object {
                            "flex": 1,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "padding": 20,
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "borderColor": "grey",
                                "borderRadius": 4,
                                "borderWidth": 0.5,
                                "padding": 15.714285714285714,
                              }
                            }
                          >
                            <TextInput
                              allowFontScaling={true}
                              onChangeText={[Function]}
                              placeholder="Search"
                              rejectResponderTermination={true}
                              style={
                                Object {
                                  "paddingVertical": 0,
                                }
                              }
                              underlineColorAndroid="transparent"
                              value=""
                            />
                          </View>
                        </View>
                        <View
                          accessibilityLabel="list"
                          style={
                            Object {
                              "paddingHorizontal": 20,
                            }
                          }
                        >
                          <View
                            accessibilityLabel="List"
                            accessible={true}
                          >
                            <RCTScrollView
                              keyboardShouldPersistTaps="handled"
                              showsVerticalScrollIndicator={false}
                            >
                              <View />
                            </RCTScrollView>
                          </View>
                        </View>
                      </View>
                    </RCTSafeAreaView>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenContainer>
  </RNCSafeAreaProvider>
</View>
`;
